### ISSUE TITLE
systemd-resolved doing a pointless DS query

### ISSUE THREAD

**Issue by bigon** [systemd/systemd#8897]
Hello,

Running `systemd-resolve www.facebook.com` hangs for 1min and then fails with `www.facebook.com: resolve call failed: DNSSEC validation failed: failed-auxiliary`. `systemd-resolve facebook.com` works without any problem. 

www.facebook.com is a CNAME to star-mini.c10r.facebook.com. while facebook.com is a A record.

This is the same issue as #3408

[log.txt](https://github.com/systemd/systemd/files/1971317/log.txt)

Edit: systemd 238 (debian unstable) `DNSSEC setting: allow-downgrade`

**Comment by shibumi** [comment]
Works fine on my machine with Arch Linux and systemd 238.76-1.

Can you provide a full resolved.conf?

**Comment by bigon** [comment]
It's the default config file:

```
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.
#
# Entries in this file show the compile time defaults.
# You can change settings by editing this file.
# Defaults can be restored by simply deleting this file.
#
# See resolved.conf(5) for details

[Resolve]
#DNS=
#FallbackDNS=8.8.8.8 8.8.4.4 2001:4860:4860::8888 2001:4860:4860::8844
#Domains=
#LLMNR=yes
#MulticastDNS=yes
#DNSSEC=allow-downgrade
#Cache=yes
#DNSStubListener=udp
 ```

**Comment by yuwata** [comment]
I cannot reproduce this.
```
$ systemd-resolve www.facebook.com
www.facebook.com: 31.13.82.36
                  (star-mini.c10r.facebook.com)

-- Information acquired via protocol DNS in 201.0ms.
-- Data is authenticated: no
```

**Comment by shibumi** [comment]
same here
```
www.facebook.com: 157.240.20.35
                  (star-mini.c10r.facebook.com)

-- Information acquired via protocol DNS in 174.5ms.
-- Data is authenticated: no
```

**Comment by bigon** [comment]
```
bigon@valinor:~$ dig www.facebook.com +dnssec 

; <<>> DiG 9.11.3-1-Debian <<>> www.facebook.com +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 40831
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 2, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1280
;; QUESTION SECTION:
;www.facebook.com.		IN	A

;; ANSWER SECTION:
www.facebook.com.	161	IN	CNAME	star-mini.c10r.facebook.com.
star-mini.c10r.facebook.com. 52	IN	A	179.60.195.36

;; AUTHORITY SECTION:
c10r.facebook.com.	600	IN	NS	a.ns.c10r.facebook.com.
c10r.facebook.com.	600	IN	NS	b.ns.c10r.facebook.com.

;; ADDITIONAL SECTION:
a.ns.c10r.facebook.com.	1115	IN	A	69.171.239.11
a.ns.c10r.facebook.com.	2934	IN	AAAA	2a03:2880:fffe:b:face:b00c:0:99
b.ns.c10r.facebook.com.	1759	IN	A	69.171.255.11
b.ns.c10r.facebook.com.	2979	IN	AAAA	2a03:2880:ffff:b:face:b00c:0:99

;; Query time: 5 msec
;; SERVER: 10.200.0.200#53(10.200.0.200)
;; WHEN: Tue May 08 17:25:18 CEST 2018
;; MSG SIZE  rcvd: 213

bigon@valinor:~$ dig facebook.com +dnssec 

; <<>> DiG 9.11.3-1-Debian <<>> facebook.com +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26404
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 2, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1280
;; QUESTION SECTION:
;facebook.com.			IN	A

;; ANSWER SECTION:
facebook.com.		133	IN	A	179.60.195.36

;; AUTHORITY SECTION:
facebook.com.		600	IN	NS	b.ns.facebook.com.
facebook.com.		600	IN	NS	a.ns.facebook.com.

;; ADDITIONAL SECTION:
a.ns.facebook.com.	70796	IN	A	69.171.239.12
a.ns.facebook.com.	146855	IN	AAAA	2a03:2880:fffe:c:face:b00c:0:35
b.ns.facebook.com.	10827	IN	A	69.171.255.12
b.ns.facebook.com.	28919	IN	AAAA	2a03:2880:ffff:c:face:b00c:0:35

;; Query time: 5 msec
;; SERVER: 10.200.0.200#53(10.200.0.200)
;; WHEN: Tue May 08 17:25:21 CEST 2018
;; MSG SIZE  rcvd: 180

bigon@valinor:~$ time systemd-resolve www.facebook.com
www.facebook.com: resolve call failed: DNSSEC validation failed: failed-auxiliary

real	1m22,919s
user	0m0,006s
sys	0m0,000s
bigon@valinor:~$ time systemd-resolve facebook.com
facebook.com: 179.60.195.36

-- Information acquired via protocol DNS in 53.7ms.
-- Data is authenticated: no

real	0m0,059s
user	0m0,005s
sys	0m0,000s
```

**Comment by bigon** [comment]
The only think I see in the logs is `Using degraded feature set (UDP+EDNS0+DO) for DNS server 10.x.x.x.` so it's not using the LARGE feature

**Comment by bigon** [comment]
```
bigon@valinor:~$ dig www.facebook.com +dnssec 

; <<>> DiG 9.11.3-1-Debian <<>> www.facebook.com +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27681
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 2, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 4096
;; QUESTION SECTION:
;www.facebook.com.		IN	A

;; ANSWER SECTION:
www.facebook.com.	1802	IN	CNAME	star-mini.c10r.facebook.com.
star-mini.c10r.facebook.com. 0	IN	A	179.60.195.36

;; AUTHORITY SECTION:
c10r.facebook.com.	549	IN	NS	a.ns.c10r.facebook.com.
c10r.facebook.com.	549	IN	NS	b.ns.c10r.facebook.com.

;; ADDITIONAL SECTION:
a.ns.c10r.facebook.com.	912	IN	A	69.171.239.11
a.ns.c10r.facebook.com.	932	IN	AAAA	2a03:2880:fffe:b:face:b00c:0:99
b.ns.c10r.facebook.com.	3438	IN	A	69.171.255.11
b.ns.c10r.facebook.com.	1649	IN	AAAA	2a03:2880:ffff:b:face:b00c:0:99

;; Query time: 2 msec
;; SERVER: 10.122.17.186#53(10.122.17.186)
;; WHEN: Tue May 08 17:58:10 CEST 2018
;; MSG SIZE  rcvd: 213

bigon@valinor:~$ time systemd-resolve www.facebook.com
www.facebook.com: 179.60.195.36
                  (star-mini.c10r.facebook.com)

-- Information acquired via protocol DNS in 22.2125s.
-- Data is authenticated: no

real	0m22,220s
user	0m0,007s
sys	0m0,001s
bigon@valinor:~$ time systemd-resolve www.facebook.com
www.facebook.com: 179.60.195.36
                  (star-mini.c10r.facebook.com)

-- Information acquired via protocol DNS in 2.5ms.
-- Data is authenticated: no

real	0m0,008s
user	0m0,000s
sys	0m0,006s
```

If I'm taking an other DNS server here, I can see that.

The difference is that this DNS server supports 4k packets according to dig and the other only 1280 bytes

**Comment by bigon** [comment]
Surprisingly I cannot reproduce that at home, I'll investigate more more why it's happening only at work

**Comment by ott** [comment]
So do you think that we can close this issue because it's not a problem with systemd-resolved? Or is it a problem of the interaction of systemd-resolved with the DNS resolvers at your workplace?

**Comment by bigon** [comment]
Yeah I think it can be closed

**Comment by bigon** [comment]
FTR: https://lists.isc.org/pipermail/bind-users/2018-September/100666.html

It seems that the issue is a mix of facebook ignoring that DNSSEC exists/bind not playing nice with that fact and resolved doing a pointless DS query(?)

**Comment by bigon** [comment]
So to summary, when doing a query to get DS record to c10r.facebook.com via a bind server, bind replies with a SERVFAIL (this is caused by an issue on the DNS of facebook and should probably be fixed there):

```
$ dig ds c10r.facebook.com @10.122.17.186

; <<>> DiG 9.11.4-P1-1-Debian <<>> ds c10r.facebook.com @10.122.17.186
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 6750
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;c10r.facebook.com.		IN	DS

;; Query time: 34 msec
;; SERVER: 10.122.17.186#53(10.122.17.186)
;; WHEN: mar sep 04 11:05:40 CEST 2018
;; MSG SIZE  rcvd: 46
```

As you can see the query is fast.

But when doing the query with systemd-resolve, it takes sometimes up to a second (which is better than in my initial bug report as we changed something on the server side, but this is still a lot)

```
$ systemd-resolve www.facebook.com
www.facebook.com: 179.60.195.38
                  (star-z-mini.c10r.facebook.com)

-- Information acquired via protocol DNS in 905.3ms.
-- Data is authenticated: no

```
There is some analysis from the logs I sent a while back: https://lists.isc.org/pipermail/bind-users/2018-September/100672.html

**Comment by bigon** [comment]
TLDR; Use the bind9 DNS servers from https://www.dns-oarc.net/oarc/services/odvr to reproduce

**Comment by chantra** [comment]
I fixed the behaviour for domains running on Facebook NS servers, so this should not be an issue anymore for FB specific names. You will still hit this issue with potentially other names out there though.

**Comment by ott** [comment]
@chantra What was the reason for the behaviour? Is the defect in systemd-resolved or the DNS server software that you were using?

**Comment by chantra** [comment]
@ott FB nameservers were treating DS record just like any other records, when it needs to be handled by the [parent of the zone](https://tools.ietf.org/html/rfc3658#section-2.2.1.1).... but systemd is also poorly handling that case.
https://lists.isc.org/pipermail/bind-users/2018-September/100672.html has more details.

**Comment by JpegXguy** [comment]
Seems like my situation is related to this issue. More here:
https://community.cloudflare.com/t/sites-that-are-aliased-to-ec-azureedge-net-cannot-be-resolved-if-dnssec-is-on/94282

**Comment by mnordhoff** [comment]
Facebook has fixed this issue, but you should still be able to reproduce it with other affected DNS services, like `www.netflix.com`.

**Comment by pemensik** [comment]
I think this has been fixed by https://github.com/systemd/systemd/pull/31827.
